templates/default/deploy-scripts/src/deployer.rs

```diff
@@ -0,0 +1,453 @@
+
+        let expiry = expiry.unwrap_or(TransactionTime::from_seconds(
+            (chrono::Utc::now().timestamp() + 300) as u64,
+        ));
+
+        let tx = init_contract(&self.key, self.key.address, nonce.nonce, expiry, payload, energy);
+
+        let bi = transactions::BlockItem::AccountTransaction(tx);
+
+        let tx_hash = self
+            .client
+            .clone()
+            .send_block_item(&bi)
+            .await
+            .map_err(DeployError::TransactionRejected)?;
+
+        println!("Sent tx: {tx_hash}");
+
+        let (_, block_item) = self.client.wait_until_finalized(&tx_hash).await?;
+
+        let contract_init = self.check_outcome_of_initialization_transaction(block_item)?;
+
+        println!(
+            "Transaction finalized, tx_hash={} contract=({}, {})",
+            tx_hash, contract_init.contract.index, contract_init.contract.subindex,
+        );
+
+        Ok((tx_hash, contract_init.contract))
+    }
+
+    /// A function to update a smart contract instance on the chain.
+    ///
+    /// If successful, the transaction
+    /// hash is returned.
+    ///
+    /// An optional energy for the transaction can be
+    /// given. If `None` is provided, 50000 energy is used as a default energy
+    /// value. An optional expiry time for the transaction can be given. If
+    /// `None` is provided, the local time + 300 seconds is used as a default
+    /// expiry time.
+    pub async fn update_contract(
+        &mut self,
+        update_payload: UpdateContractPayload,
+        energy: Option<GivenEnergy>,
+        expiry: Option<TransactionTime>,
+    ) -> Result<HashBytes<TransactionMarker>, DeployError> {
+        println!("\nUpdating contract....");
+
+        let nonce = self.get_nonce(self.key.address).await?;
+
+        if !nonce.all_final {
+            return Err(DeployError::NonceNotFinal);
+        }
+
+        let payload = transactions::Payload::Update {
+            payload: update_payload,
+        };
+
+        let expiry = expiry.unwrap_or(TransactionTime::from_seconds(
+            (chrono::Utc::now().timestamp() + 300) as u64,
+        ));
+
+        let energy = energy.unwrap_or(GivenEnergy::Absolute(Energy {
+            energy: 50000,
+        }));
+
+        let tx = transactions::send::make_and_sign_transaction(
+            &self.key,
+            self.key.address,
+            nonce.nonce,
+            expiry,
+            energy,
+            payload,
+        );
+        let bi = transactions::BlockItem::AccountTransaction(tx);
+
+        let tx_hash = self
+            .client
+            .clone()
+            .send_block_item(&bi)
+            .await
+            .map_err(DeployError::TransactionRejected)?;
+
+        println!("Sent tx: {tx_hash}");
+
+        let (_, block_item) = self.client.wait_until_finalized(&tx_hash).await?;
+
+        self.check_outcome_of_update_transaction(block_item)?;
+
+        println!("Transaction finalized, tx_hash={}", tx_hash,);
```

Comment:
I don't like that these fucntions both print and return stuff. We should probably just return values and print them in main. And this module is just responsible for doing things.

Reply:
In my opinion, printing these statements should be the default behavior. We don't want every developer to have to add their own custom "logging/printing" to the deployment script in every script.

Logging/Printing is very helpful in case the script fails to execute successfully, then the script can be easily re-run from the point it failed (fixing the bug that caused it to fail beforehand of course).

I suggest to add a flag to "switch off" the verbose logging in case somebody wants to write their custom logging.

Reply:
I suggest adding logging to the functions that execute a transaction on chain (meaning `deploy`/`initialize`/`update` transactions). All other functions (read functions e.g. estimate energy/ get nonce) should not log any output.

Reply:
Ok, then keep it simple without switches to turn it off.

If it is meant to log then make sure to log consistently before every action that might fail.

Reply:
I added already a commit that adds the flag to switch off logging yesterday.

Reply:
But why is it sensible to have this option? Doesn't it just complicate the code?

As I see it this is not a library, this is the deployment script for this particular contract.  And you argued that we should just log things since it's easier to debug and users don't have to recreate it every time. That makes sense to me.

But then adding this flag seems to me to go against this simplicity of use. If this was a library that we publish then I do agree we should have logging configurable, but then we should use the `log` crate for it since that is the standard way of doing logging in Rust, and allows the caller to configure it, enable it or disable it.

But since this is not a library, it is the template for the code they'll write it seems unnecessary complexity to add a flag like this.

Reply:
Removed logging flag. Logging is done on the `deploy/initialize/update` transactions.